content/blog/2023/2023-09-16-create-openfeign-http-client.md

```diff
@@ -0,0 +1,503 @@
+```java
+public class CalculatorErrorDecoder implements ErrorDecoder {
+  private final ErrorDecoder defaultErrorDecoder = new Default();
+
+  @Override
+  public Exception decode(String methodKey, Response response) {
+    ExceptionMessage message = null;
+    try (InputStream bodyIs = response.body().asInputStream()) {
+      ObjectMapper mapper = new ObjectMapper();
+      message = mapper.readValue(bodyIs, ExceptionMessage.class);
+    } catch (IOException e) {
+      return new Exception(e.getMessage());
+    }
+    final String messageStr = message == null ? "" : message.getMessage();
+    switch (response.status()) {
+      case 400:
+        return new RuntimeException(messageStr.isEmpty() ? "Bad Request" : messageStr);
+      case 401:
+        return new RetryableException(response.status(),
+                                      response.reason(),
+                                      response.request().httpMethod(),
+                                      null,
+                                      response.request());
+      case 404:
+        return new RuntimeException(messageStr.isEmpty() ? "Not found" : messageStr);
+      default:
+        return defaultErrorDecoder.decode(methodKey, response);
+    }
+  }
+}
+```
+All responses with HTTP status other than `HTTP 2xx` range, for example `HTTP 400`, will trigger the ErrorDecoder's `decode` method. 
+In this overridden `decode` method, we can handle the response, wrap the failure into a custom exception or perform any additional processing.
+
+
+We can even retry the request again by throwing a `RetryableException`. 
+This will invoke the registered `Retryer`. [Retryer](#configuring-retryer) is explained in detail in the advanced techniques.
+
+You can see it in action by running `givenNegativeDivisorDivisionReturnsError` test in the example code shared on [Github](#example-code).
+
+## Advanced Techniques
+
+### Integrating Encoder/Decoder
+Encoder and decoder are **used to encode/decode the request and response data respectively**. We select these depending on the content type of the request and response. For example, [Gson](https://github.com/OpenFeign/feign/blob/master/gson) or [Jackson](https://github.com/OpenFeign/feign/blob/master/jackson)
+ can be used for JSON data.
+
+Here is an example showing how to use `Jackson` encoder and decoder.
+```java
+final CalculatorService target = Feign.builder()
+                                      .encoder(new JacksonEncoder())
+                                      .decoder(new JacksonDecoder())
+                                      .target(CalculatorService.class, HOST);
+
+```
+
+### Changing HTTP Client
+By default, it uses Feign HTTP client. The motivation behind changing the default HTTP client of Feign, from the original Apache HTTP Client to other libraries like OkHttp, is primarily driven by the need for **better performance, improved features, and enhanced compatibility with modern HTTP standards**.
```

Comment:
> By default, it uses Feign HTTP client

This will lead to confusion. Aren't you explaining a Feign client in this article? I know that you mean the underlying HTTP client, but some readers won't know that, because you haven't explained it. Please add another sentence or so explaining that Feign is "just" a declarative layer and what the difference is to a "real" HTTP client.